### PR TITLE
Make sure the index array is contiguous before changing dtype

### PR DESCRIPTION
**Pull request by lamblin**
I'm not completely sure about the incref/decref interaction with Py_BuildValue and PyArg_ParseTuple, though.

### CODE REVIEW

**Comment by mohammadpz** [comment]
I get this error:
```python
*** glibc detected *** python: invalid fastbin entry (free): 0x000000000e568ea0 ***
======= Backtrace: =========
/lib64/libc.so.6[0x3e30a76d63]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(+0x6506a)[0x7fbf4ae7106a]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(+0x66f57)[0x7fbf4ae72f57]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(+0x98aa8)[0x7fbf4aea4aa8]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyObject_Call+0x68)[0x7fbf4ae52388]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x5b27)[0x7fbf4aee8a27]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x54ff)[0x7fbf4aee83ff]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x54ff)[0x7fbf4aee83ff]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x54ff)[0x7fbf4aee83ff]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x54ff)[0x7fbf4aee83ff]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x54ff)[0x7fbf4aee83ff]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x54ff)[0x7fbf4aee83ff]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x54ff)[0x7fbf4aee83ff]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x6025)[0x7fbf4aee8f25]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(+0x6d202)[0x7fbf4ae79202]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyObject_Call+0x68)[0x7fbf4ae52388]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalFrameEx+0xd29)[0x7fbf4aee3c29]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x8d2)[0x7fbf4aeea4c2]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyEval_EvalCode+0x32)[0x7fbf4aeea512]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(+0xf8182)[0x7fbf4af04182]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyRun_FileExFlags+0x96)[0x7fbf4af04256]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(PyRun_SimpleFileExFlags+0x1c9)[0x7fbf4af057e9]
/part/02/Tmp/lisa/os_v3/canopy/bin/../lib/libpython2.7.so.1.0(Py_Main+0xc96)[0x7fbf4af16736]
python(main+0x3e5)[0x4012c5]
/lib64/libc.so.6(__libc_start_main+0xfd)[0x3e30a1ee5d]
python[0x400cd9]
======= Memory map: ========
00400000-00402000 r-xp 00000000 08:11 231348861                          /part/02/Tmp/lisa/os_v3/canopy/bin/python
00601000-00602000 rw-p 00001000 08:11 231348861                          /part/02/Tmp/lisa/os_v3/canopy/bin/python
00fbe000-285c5000 rw-p 00000000 00:00 0                                  [heap]
200000000-200100000 rw-s 4fbfb9000 00:05 11968                           /dev/nvidiactl
200100000-203100000 ---p 00000000 00:00 0 
203100000-203200000 rw-s 509ff1000 00:05 11968                           /dev/nvidiactl
203200000-203300000 ---p 00000000 00:00 0 
203300000-203400000 rw-s 42a3d9000 00:05 11968                           /dev/nvidiactl
203400000-1100000000 ---p 00000000 00:00 0 
3ba3c00000-3ba3c04000 r-xp 00000000 08:02 14310                          /lib64/libuuid.so.1.3.0
3ba3c04000-3ba3e03000 ---p 00004000 08:02 14310                          /lib64/libuuid.so.1.3.0
3ba3e03000-3ba3e04000 rw-p 00003000 08:02 14310                          /lib64/libuuid.so.1.3.0
3e30600000-3e3061f000 r-xp 00000000 08:02 14041                          /lib64/ld-2.13.so
3e3081e000-3e3081f000 r--p 0001e000 08:02 14041                          /lib64/ld-2.13.so
3e3081f000-3e30820000 rw-p 0001f000 08:02 14041                          /lib64/ld-2.13.so
3e30820000-3e30821000 rw-p 00000000 00:00 0 
3e30a00000-3e30b91000 r-xp 00000000 08:02 13959                          /lib64/libc-2.13.so
3e30b91000-3e30d91000 ---p 00191000 08:02 13959                          /lib64/libc-2.13.so
3e30d91000-3e30d95000 r--p 00191000 08:02 13959                          /lib64/libc-2.13.so
3e30d95000-3e30d96000 rw-p 00195000 08:02 13959                          /lib64/libc-2.13.so
3e30d96000-3e30d9c000 rw-p 00000000 00:00 0 
3e30e00000-3e30e84000 r-xp 00000000 08:02 13984                          /lib64/libm-2.13.so
3e30e84000-3e31083000 ---p 00084000 08:02 13984                          /lib64/libm-2.13.so
3e31083000-3e31084000 r--p 00083000 08:02 13984                          /lib64/libm-2.13.so
3e31084000-3e31085000 rw-p 00084000 08:02 13984                          /lib64/libm-2.13.so
3e31200000-3e31202000 r-xp 00000000 08:02 14053                          /lib64/libdl-2.13.so
3e31202000-3e31402000 ---p 00002000 08:02 14053                          /lib64/libdl-2.13.so
3e31402000-3e31403000 r--p 00002000 08:02 14053                          /lib64/libdl-2.13.so
3e31403000-3e31404000 rw-p 00003000 08:02 14053                          /lib64/libdl-2.13.so
3e31600000-3e31617000 r-xp 00000000 08:02 14005                          /lib64/libpthread-2.13.so
3e31617000-3e31816000 ---p 00017000 08:02 14005                          /lib64/libpthread-2.13.so
3e31816000-3e31817000 r--p 00016000 08:02 14005                          /lib64/libpthread-2.13.so
3e31817000-3e31818000 rw-p 00017000 08:02 14005                          /lib64/libpthread-2.13.so
3e31818000-3e3181c000 rw-p 00000000 00:00 0 
3e31a00000-3e31a15000 r-xp 00000000 08:02 13966                          /lib64/libgcc_s-4.5.1-20100924.so.1
3e31a15000-3e31c14000 ---p 00015000 08:02 13966                          /lib64/libgcc_s-4.5.1-20100924.so.1
3e31c14000-3e31c15000 rw-p 00014000 08:02 13966                          /lib64/libgcc_s-4.5.1-20100924.so.1
3e32200000-3e32207000 r-xp 00000000 08:02 14007                          /lib64/librt-2.13.so
3e32207000-3e32406000 ---p 00007000 08:02 14007                          /lib64/librt-2.13.so
3e32406000-3e32407000 r--p 00006000 08:02 14007                          /lib64/librt-2.13.so
3e32407000-3e32408000 rw-p 00007000 08:02 14007                          /lib64/librt-2.13.so
3e34e00000-3e34ee8000 r-xp 00000000 08:02 128894                         /usr/lib64/libstdc++.so.6.0.14
3e34ee8000-3e350e8000 ---p 000e8000 08:02 128894                         /usr/lib64/libstdc++.so.6.0.14
3e350e8000-3e350f0000 r--p 000e8000 08:02 128894                         /usr/lib64/libstdc++.so.6.0.14
3e350f0000-3e350f2000 rw-p 000f0000 08:02 128894                         /usr/lib64/libstdc++.so.6.0.14
3e350f2000-3e35107000 rw-p 00000000 00:00 0 
3e46200000-3e46203000 r-xp 00000000 08:02 13975                          /lib64/libgpg-error.so.0.7.0
3e46203000-3e46402000 ---p 00003000 08:02 13975                          /lib64/libgpg-error.so.0.7.0
3e46402000-3e46403000 rw-p 00002000 08:02 13975                          /lib64/libgpg-error.so.0.7.0
3e46600000-3e46602000 r-xp 00000000 08:02 14314                          /lib64/libutil-2.13.so
3e46602000-3e46801000 ---p 00002000 08:02 14314                          /lib64/libutil-2.13.so
3e46801000-3e46802000 r--p 00001000 08:02 14314                          /lib64/libutil-2.13.so
3e46802000-3e46803000 rw-p 00002000 08:02 14314                          /lib64/libutil-2.13.so
3e47600000-3e47672000 r-xp 00000000 08:02 13972                          /lib64/libgcrypt.so.11.5.3
3e47672000-3e47871000 ---p 00072000 08:02 13972                          /lib64/libgcrypt.so.11.5.3
3e47871000-3e47875000 rw-p 00071000 08:02 13972                          /lib64/libgcrypt.so.11.5.3
7fbeff2d0000-7fbeffc11000 rw-p 00000000 00:00 0 
7fbeffc11000-7fbeffc18000 r-xp 00000000 08:02 528732                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpUyH5Ij/3587c797dbdc0159676187078a7ba405.so
7fbeffc18000-7fbeffe17000 ---p 00007000 08:02 528732                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpUyH5Ij/3587c797dbdc0159676187078a7ba405.so
7fbeffe17000-7fbeffe18000 rw-p 00006000 08:02 528732                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpUyH5Ij/3587c797dbdc0159676187078a7ba405.so
7fbeffe18000-7fbeffe1d000 r-xp 00000000 08:02 528726                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmplXDYIk/085514dd9cca93fb228c601396f93810.so
7fbeffe1d000-7fbf0001d000 ---p 00005000 08:02 528726                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmplXDYIk/085514dd9cca93fb228c601396f93810.so
7fbf0001d000-7fbf0001e000 rw-p 00005000 08:02 528726                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmplXDYIk/085514dd9cca93fb228c601396f93810.so
7fbf0001e000-7fbf00024000 r-xp 00000000 08:02 528719                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpnTTong/52066f5dcfaf44ab89bb80152e8303cb.so
7fbf00024000-7fbf00224000 ---p 00006000 08:02 528719                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpnTTong/52066f5dcfaf44ab89bb80152e8303cb.so
7fbf00224000-7fbf00225000 rw-p 00006000 08:02 528719                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpnTTong/52066f5dcfaf44ab89bb80152e8303cb.so
7fbf00225000-7fbf00228000 r-xp 00000000 08:02 528713                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpyCRpZV/480ff53bc9ed2b78a08898f3c74bb43d.so
7fbf00228000-7fbf00427000 ---p 00003000 08:02 528713                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpyCRpZV/480ff53bc9ed2b78a08898f3c74bb43d.so
7fbf00427000-7fbf00428000 rw-p 00002000 08:02 528713                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpyCRpZV/480ff53bc9ed2b78a08898f3c74bb43d.so
7fbf00428000-7fbf00430000 r-xp 00000000 08:02 528706                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpX6S1IC/c9513b3258a826fa85449a7dcc40c8c1.so
7fbf00430000-7fbf00630000 ---p 00008000 08:02 528706                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpX6S1IC/c9513b3258a826fa85449a7dcc40c8c1.so
7fbf00630000-7fbf00631000 rw-p 00008000 08:02 528706                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpX6S1IC/c9513b3258a826fa85449a7dcc40c8c1.so
7fbf00631000-7fbf00638000 r-xp 00000000 08:02 528700                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpicYZ36/a7014ccdeb99079a05fe5f45d4b13e27.so
7fbf00638000-7fbf00838000 ---p 00007000 08:02 528700                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpicYZ36/a7014ccdeb99079a05fe5f45d4b13e27.so
7fbf00838000-7fbf00839000 rw-p 00007000 08:02 528700                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpicYZ36/a7014ccdeb99079a05fe5f45d4b13e27.so
7fbf00839000-7fbf00841000 r-xp 00000000 08:02 528738                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmp6UjGgo/5d53f90fb9614696352dbd44fb4d44ef.so
7fbf00841000-7fbf00a40000 ---p 00008000 08:02 528738                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmp6UjGgo/5d53f90fb9614696352dbd44fb4d44ef.so
7fbf00a40000-7fbf00a41000 rw-p 00007000 08:02 528738                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmp6UjGgo/5d53f90fb9614696352dbd44fb4d44ef.so
7fbf00a41000-7fbf00a44000 r-xp 00000000 08:02 528687                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpj6A5pd/02066d195ff330e320f3b0caf54a3885.so
7fbf00a44000-7fbf00c43000 ---p 00003000 08:02 528687                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpj6A5pd/02066d195ff330e320f3b0caf54a3885.so
7fbf00c43000-7fbf00c44000 rw-p 00002000 08:02 528687                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpj6A5pd/02066d195ff330e320f3b0caf54a3885.so
7fbf00c44000-7fbf00c4c000 r-xp 00000000 08:02 528681                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpIQyfc2/e1781c963c87943eff7e2d31e4c15bce.so
7fbf00c4c000-7fbf00e4b000 ---p 00008000 08:02 528681                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpIQyfc2/e1781c963c87943eff7e2d31e4c15bce.so
7fbf00e4b000-7fbf00e4c000 rw-p 00007000 08:02 528681                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpIQyfc2/e1781c963c87943eff7e2d31e4c15bce.so
7fbf00e4c000-7fbf00e52000 r-xp 00000000 08:02 528675                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpI0pxi_/670dfeb679fa019adcf10cd27dcf2602.so
7fbf00e52000-7fbf01052000 ---p 00006000 08:02 528675                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpI0pxi_/670dfeb679fa019adcf10cd27dcf2602.so
7fbf01052000-7fbf01053000 rw-p 00006000 08:02 528675                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmpI0pxi_/670dfeb679fa019adcf10cd27dcf2602.so
7fbf01053000-7fbf01057000 r-xp 00000000 08:02 528668                     /tmp/pezeshki/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.6-64/tmp4plDwl/bc35ef9c0efe8ac50115f91cac712cf5.soAborted```

File: theano/sandbox/cuda/cuda_ndarray.cu [hunk 1]
```diff
@@ -1026,13 +1026,21 @@ CudaNdarray_TakeFrom(CudaNdarray * self, PyObject *args){
                             " indices with only 1 dimensions");
             return NULL;
         }
+        // We need indices_obj to be contiguous, in order to take a view
+        // with a different dtype.
+        if (!PyArray_IS_C_CONTIGUOUS((PyArrayObject*) indices_obj)) {
```
Comment (nouiz) [inline theano/sandbox/cuda/cuda_ndarray.cu]: I would suggest that you use: tmp = PyArray_GETCONTIGUOUS(indices_obj)

Then always decref tmp when you don't need it anymore.

Comment (lamblin) [inline theano/sandbox/cuda/cuda_ndarray.cu]: Yes, I'll do that.

**Comment by lamblin** [comment]
@mohammadpz Sorry for the delay, hopefully the last push fixes it. Could you please test it?

**Comment by mohammadpz** [comment]
@lamblin ,
Awesome! It's working now. Thanks a lot Pascal :+1:

**Comment by nouiz** [comment]
Seem good. Good finding.

**Pull request merged**